File: src/topo_shadow_box/core/elevation.py
"""Elevation data fetching from AWS Terrain-RGB tiles (Terrarium format)."""

import math
import numpy as np
from io import BytesIO

import httpx
from PIL import Image
from scipy import interpolate
from scipy.ndimage import gaussian_filter

# AWS Terrain Tiles (Mapzen/Tilezen Terrarium format) - free, globally available
AWS_TERRAIN_URL = "https://s3.amazonaws.com/elevation-tiles-prod/terrarium/{z}/{x}/{y}.png"


def _lat_lon_to_tile(lat: float, lon: float, zoom: int) -> tuple[int, int]:
    """Convert lat/lon to tile coordinates at a given zoom level."""
    lat_rad = math.radians(lat)
    n = 2.0 ** zoom
    x = int((lon + 180.0) / 360.0 * n)
    y = int((1.0 - math.asinh(math.tan(lat_rad)) / math.pi) / 2.0 * n)
    return x, y


def _tile_to_lat_lon(x: int, y: int, zoom: int) -> tuple[float, float]:
    """Convert tile coordinates to lat/lon (northwest corner)."""
    n = 2.0 ** zoom
    lon = x / n * 360.0 - 180.0
    lat_rad = math.atan(math.sinh(math.pi * (1 - 2 * y / n)))
    lat = math.degrees(lat_rad)
    return lat, lon


def _decode_terrarium(r: np.ndarray, g: np.ndarray, b: np.ndarray) -> np.ndarray:
    """Decode Terrarium format RGB to elevation in meters."""
    return (r * 256.0 + g + b / 256.0) - 32768.0


def _pick_zoom(north: float, south: float, east: float, west: float) -> int:
    """Pick an appropriate zoom level based on area span."""
    max_span = max(north - south, east - west)
    if max_span > 1.0:
        return 10
    elif max_span > 0.5:
        return 11
    elif max_span > 0.1:
        return 12
    elif max_span > 0.05:
        return 13
    else:
        return 14


async def fetch_terrain_elevation(
    north: float, south: float, east: float, west: float,
    resolution: int = 200,
) -> dict:
    """Fetch elevation data for a bounding box from AWS Terrain-RGB tiles.

    Returns dict with: grid (ndarray), lats (ndarray), lons (ndarray),
    min_elevation, max_elevation
    """
    zoom = _pick_zoom(north, south, east, west)

    x_min, y_max = _lat_lon_to_tile(south, west, zoom)
    x_max, y_min = _lat_lon_to_tile(north, east, zoom)

    if x_min > x_max:
        x_min, x_max = x_max, x_min
    if y_min > y_max:
        y_min, y_max = y_max, y_min

    num_tiles_x = x_max - x_min + 1
    num_tiles_y = y_max - y_min + 1

    if num_tiles_x * num_tiles_y > 25:
        # Fall back to lower zoom
        zoom = max(zoom - 1, 8)
        x_min, y_max = _lat_lon_to_tile(south, west, zoom)
        x_max, y_min = _lat_lon_to_tile(north, east, zoom)
        if x_min > x_max:
            x_min, x_max = x_max, x_min
        if y_min > y_max:
            y_min, y_max = y_max, y_min
        num_tiles_x = x_max - x_min + 1
        num_tiles_y = y_max - y_min + 1

    tile_size = 256
    stitched_width = num_tiles_x * tile_size
    stitched_height = num_tiles_y * tile_size
    stitched_elevations = np.zeros((stitched_height, stitched_width))

    async with httpx.AsyncClient(timeout=30.0) as client:
        for ty in range(y_min, y_max + 1):
            for tx in range(x_min, x_max + 1):
                url = AWS_TERRAIN_URL.format(z=zoom, x=tx, y=ty)
                try:
                    response = await client.get(url)
                    if response.status_code == 200:
                        img = Image.open(BytesIO(response.content))
                        img_array = np.array(img)
                        r = img_array[:, :, 0].astype(np.float64)
                        g = img_array[:, :, 1].astype(np.float64)
                        b = img_array[:, :, 2].astype(np.float64)
                        tile_elevations = _decode_terrarium(r, g, b)

                        px = (tx - x_min) * tile_size
                        py = (ty - y_min) * tile_size
                        stitched_elevations[py:py + tile_size, px:px + tile_size] = tile_elevations
                except Exception:
                    pass  # Leave zeros for failed tiles

    # Map tile pixel coordinates to geographic coordinates
    tile_north, tile_west = _tile_to_lat_lon(x_min, y_min, zoom)
    tile_south, tile_east = _tile_to_lat_lon(x_max + 1, y_max + 1, zoom)

    tile_lats = np.linspace(tile_south, tile_north, stitched_height)
    tile_lons = np.linspace(tile_west, tile_east, stitched_width)

    # Flip so that index 0 = south, last = north
    stitched_elevations = np.flipud(stitched_elevations)

    # Crop to requested bounding box
    lat_indices = np.where((tile_lats >= south) & (tile_lats <= north))[0]
    lon_indices = np.where((tile_lons >= west) & (tile_lons <= east))[0]

    if len(lat_indices) == 0 or len(lon_indices) == 0:
        raise ValueError("Bounding box doesn't overlap with fetched tiles")

    cropped = stitched_elevations[
        lat_indices[0]:lat_indices[-1] + 1,
        lon_indices[0]:lon_indices[-1] + 1,
    ]
    cropped_lats = tile_lats[lat_indices[0]:lat_indices[-1] + 1]
    cropped_lons = tile_lons[lon_indices[0]:lon_indices[-1] + 1]

    if cropped.shape[0] < 4 or cropped.shape[1] < 4:
        raise ValueError("Cropped region too small for interpolation")

    # Interpolate to requested resolution
    interp_func = interpolate.RectBivariateSpline(
        cropped_lats, cropped_lons, cropped, kx=3, ky=3,
    )
    target_lats = np.linspace(south, north, resolution)
    target_lons = np.linspace(west, east, resolution)
    elevations = interp_func(target_lats, target_lons)

    # Smooth to reduce tile boundary artifacts
    elevations = gaussian_filter(elevations, sigma=0.5)

    return {
        "grid": elevations,
        "lats": target_lats,
        "lons": target_lons,
        "min_elevation": float(np.min(elevations)),
        "max_elevation": float(np.max(elevations)),
    }

File: go/internal/core/elevation.go
package core

import (
	"context"
	"fmt"
	"image"
	_ "image/png"
	"io"
	"math"
	"net/http"
	"sync"

	"github.com/huslage/topo-shadow-box/internal/session"
)

const awsTerrainURL = "https://s3.amazonaws.com/elevation-tiles-prod/terrarium/%d/%d/%d.png"

// HTTPClient interface for testability
type HTTPClient interface {
	Get(url string) (*http.Response, error)
}

func LatLonToTile(lat, lon float64, zoom int) (x, y int) {
	n := math.Pow(2, float64(zoom))
	x = int((lon + 180.0) / 360.0 * n)
	latRad := lat * math.Pi / 180
	y = int((1.0 - math.Log(math.Tan(latRad)+1/math.Cos(latRad))/math.Pi) / 2.0 * n)
	return
}

func TileToLatLon(x, y, zoom int) (lat, lon float64) {
	n := math.Pow(2, float64(zoom))
	lon = float64(x)/n*360.0 - 180.0
	latRad := math.Atan(math.Sinh(math.Pi * (1 - 2*float64(y)/n)))
	lat = latRad * 180 / math.Pi
	return
}

func DecodeTerrariumPixel(r, g, b uint8) float64 {
	return float64(r)*256.0 + float64(g) + float64(b)/256.0 - 32768.0
}

func PickZoom(north, south, east, west float64) int {
	maxSpan := math.Max(north-south, east-west)
	switch {
	case maxSpan > 1.0:
		return 10
	case maxSpan > 0.5:
		return 11
	case maxSpan > 0.1:
		return 12
	case maxSpan > 0.05:
		return 13
	default:
		return 14
	}
}

func BilinearInterp(grid [][]float64, lats, lons []float64, lat, lon float64) float64 {
	rows := len(lats)
	cols := len(lons)
	if rows < 2 || cols < 2 {
		return 0
	}

	latStep := (lats[rows-1] - lats[0]) / float64(rows-1)
	lonStep := (lons[cols-1] - lons[0]) / float64(cols-1)

	r0 := int((lat - lats[0]) / latStep)
	c0 := int((lon - lons[0]) / lonStep)

	r0 = clampInt(r0, 0, rows-2)
	c0 = clampInt(c0, 0, cols-2)
	r1, c1 := r0+1, c0+1

	tLat := (lat - lats[r0]) / (lats[r1] - lats[r0])
	tLon := (lon - lons[c0]) / (lons[c1] - lons[c0])

	return (1-tLat)*(1-tLon)*grid[r0][c0] +
		(1-tLat)*tLon*grid[r0][c1] +
		tLat*(1-tLon)*grid[r1][c0] +
		tLat*tLon*grid[r1][c1]
}

func clampInt(v, lo, hi int) int {
	if v < lo {
		return lo
	}
	if v > hi {
		return hi
	}
	return v
}

// gaussianFilter3x3 applies a mild 3x3 Gaussian blur (sigma~0.5)
func gaussianFilter3x3(grid [][]float64) [][]float64 {
	k := [3][3]float64{
		{0.0751, 0.1238, 0.0751},
		{0.1238, 0.2042, 0.1238},
		{0.0751, 0.1238, 0.0751},
	}
	rows := len(grid)
	if rows == 0 {
		return grid
	}
	cols := len(grid[0])
	out := make([][]float64, rows)
	for i := range out {
		out[i] = make([]float64, cols)
	}
	for r := 1; r < rows-1; r++ {
		for c := 1; c < cols-1; c++ {
			var v float64
			for dr := -1; dr <= 1; dr++ {
				for dc := -1; dc <= 1; dc++ {
					v += k[dr+1][dc+1] * grid[r+dr][c+dc]
				}
			}
			out[r][c] = v
		}
	}
	// Copy edges unchanged
	for r := 0; r < rows; r++ {
		out[r][0] = grid[r][0]
		out[r][cols-1] = grid[r][cols-1]
	}
	for c := 0; c < cols; c++ {
		out[0][c] = grid[0][c]
		out[rows-1][c] = grid[rows-1][c]
	}
	return out
}

func linspace(start, end float64, n int) []float64 {
	if n <= 1 {
		return []float64{start}
	}
	result := make([]float64, n)
	for i := range result {
		result[i] = start + float64(i)*(end-start)/float64(n-1)
	}
	return result
}

func flipUD(grid [][]float64) {
	n := len(grid)
	for i := 0; i < n/2; i++ {
		grid[i], grid[n-1-i] = grid[n-1-i], grid[i]
	}
}

func reverseFloat64(s []float64) {
	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
		s[i], s[j] = s[j], s[i]
	}
}

// findRange returns the start and end indices in sorted slice s that cover [lo, hi].
func findRange(s []float64, lo, hi float64) (start, end int) {
	start = -1
	end = -1
	for i, v := range s {
		if v >= lo && start < 0 {
			start = i
		}
		if v <= hi {
			end = i
		}
	}
	return
}

func cropGrid(grid [][]float64, rStart, rEnd, cStart, cEnd int) [][]float64 {
	rows := rEnd - rStart + 1
	cols := cEnd - cStart + 1
	if rows <= 0 || cols <= 0 {
		return nil
	}
	out := make([][]float64, rows)
	for i := range out {
		out[i] = make([]float64, cols)
		copy(out[i], grid[rStart+i][cStart:cEnd+1])
	}
	return out
}

func gridMinMax(grid [][]float64) (min, max float64) {
	min = math.MaxFloat64
	max = -math.MaxFloat64
	for _, row := range grid {
		for _, v := range row {
			if v < min {
				min = v
			}
			if v > max {
				max = v
			}
		}
	}
	return
}

func decodeTileInto(body io.Reader, stitched [][]float64, tileX, tileY, tileSize int) {
	img, _, err := image.Decode(body)
	if err != nil {
		return
	}
	for py := 0; py < tileSize && py < img.Bounds().Dy(); py++ {
		for px := 0; px < tileSize && px < img.Bounds().Dx(); px++ {
			c := img.At(px, py)
			r, g, b, _ := c.RGBA()
			elev := DecodeTerrariumPixel(uint8(r>>8), uint8(g>>8), uint8(b>>8))
			rowIdx := tileY*tileSize + py
			colIdx := tileX*tileSize + px
			if rowIdx < len(stitched) && colIdx < len(stitched[rowIdx]) {
				stitched[rowIdx][colIdx] = elev
			}
		}
	}
}

func FetchTerrainElevation(ctx context.Context, client HTTPClient, north, south, east, west float64, resolution int) (*session.ElevationData, error) {
	if client == nil {
		client = http.DefaultClient
	}
	zoom := PickZoom(north, south, east, west)

	xMin, yMax := LatLonToTile(south, west, zoom)
	xMax, yMin := LatLonToTile(north, east, zoom)
	if xMin > xMax {
		xMin, xMax = xMax, xMin
	}
	if yMin > yMax {
		yMin, yMax = yMax, yMin
	}

	numX := xMax - xMin + 1
	numY := yMax - yMin + 1
	if numX*numY > 25 {
		if zoom > 8 {
			zoom--
		}
		xMin, yMax = LatLonToTile(south, west, zoom)
		xMax, yMin = LatLonToTile(north, east, zoom)
		if xMin > xMax {
			xMin, xMax = xMax, xMin
		}
		if yMin > yMax {
			yMin, yMax = yMax, yMin
		}
		numX = xMax - xMin + 1
		numY = yMax - yMin + 1
	}

	const tileSize = 256
	stitchedH := numY * tileSize
	stitchedW := numX * tileSize
	stitched := make([][]float64, stitchedH)
	for i := range stitched {
		stitched[i] = make([]float64, stitchedW)
	}

	var wg sync.WaitGroup
	for ty := yMin; ty <= yMax; ty++ {
		for tx := xMin; tx <= xMax; tx++ {
			wg.Add(1)
			go func(tx, ty int) {
				defer wg.Done()
				url := fmt.Sprintf(awsTerrainURL, zoom, tx, ty)
				resp, err := client.Get(url)
				if err != nil || resp.StatusCode != 200 {
					return
				}
				defer resp.Body.Close()
				decodeTileInto(resp.Body, stitched, tx-xMin, ty-yMin, tileSize)
			}(tx, ty)
		}
	}
	wg.Wait()

	tileNorth, tileWest := TileToLatLon(xMin, yMin, zoom)
	tileSouth, tileEast := TileToLatLon(xMax+1, yMax+1, zoom)

	stitchedLats := linspace(tileSouth, tileNorth, stitchedH)
	stitchedLons := linspace(tileWest, tileEast, stitchedW)

	flipUD(stitched)
	reverseFloat64(stitchedLats)

	rStart, rEnd := findRange(stitchedLats, south, north)
	cStart, cEnd := findRange(stitchedLons, west, east)
	if rStart < 0 || cStart < 0 {
		return nil, fmt.Errorf("bounding box doesn't overlap with fetched tiles")
	}
	cropped := cropGrid(stitched, rStart, rEnd, cStart, cEnd)
	croppedLats := stitchedLats[rStart : rEnd+1]
	croppedLons := stitchedLons[cStart : cEnd+1]

	if len(croppedLats) < 4 || len(croppedLons) < 4 {
		return nil, fmt.Errorf("cropped region too small for interpolation")
	}

	targetLats := linspace(south, north, resolution)
	targetLons := linspace(west, east, resolution)
	resampled := make([][]float64, resolution)
	for r := range resampled {
		resampled[r] = make([]float64, resolution)
		for c := range resampled[r] {
			resampled[r][c] = BilinearInterp(cropped, croppedLats, croppedLons, targetLats[r], targetLons[c])
		}
	}

	resampled = gaussianFilter3x3(resampled)

	minElev, maxElev := gridMinMax(resampled)
	return &session.ElevationData{
		Grid:         resampled,
		Lats:         targetLats,
		Lons:         targetLons,
		Resolution:   resolution,
		MinElevation: minElev,
		MaxElevation: maxElev,
		IsSet:        true,
	}, nil
}
